### PR TITLE
Geomap: configure size and color with different fields

### PR DESCRIPTION
**Pull request by ryantxu**
This introduces the concept of 'dimensions' that can be selected either from a fixed value or from a field value.  We use this pattern to set the fields for:

- Marker: size
- Marker: color
- Heatmap: weight

image
image
image

### CODE REVIEW

File: public/app/plugins/panel/geomap/dims/editors/ColorDimensionEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import React, { FC, useCallback } from 'react';
+import { GrafanaTheme2, SelectableValue, StandardEditorProps } from '@grafana/data';
+import { ColorDimensionConfig } from '../types';
+import { Select, ColorPicker, useStyles2 } from '@grafana/ui';
+import {
+  useFieldDisplayNames,
+  useSelectOptions,
+} from '../../../../../../../packages/grafana-ui/src/components/MatchersUI/utils';
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/dims/editors/ColorDimensionEditor.tsx:8]: typically this is bad... but I do not yet want to export them from `@grafana/ui`

File: public/app/plugins/panel/geomap/layers/data/heatMap.tsx [hunk 2]
```diff
@@ -77,47 +83,55 @@ export const heatmapLayer: MapLayerRegistryItem<HeatmapConfig> = {
...
-        } else {
-          // Set the gradient back to default if threshold or single color is chosen
-          vectorLayer.setGradient(['#00f', '#0ff', '#0f0', '#ff0', '#f00']);
+        // Set heatmap gradient colors
+        let colors = ['#00f', '#0ff', '#0f0', '#ff0', '#f00'];
+        let field = weightDim.field;
+        if(!field) { // Use the first nummeric field color scheme
+          field = frame.fields.find(field => field.type === FieldType.number);
         }
+        if(field) {
```
Comment (lean257) [inline public/app/plugins/panel/geomap/layers/data/heatMap.tsx:104]: else?

**Review by lean257 — APPROVED** [review]

File: public/app/plugins/panel/geomap/layers/data/heatMap.tsx (hunk 2 above)
Comment (ryantxu) [reply public/app/plugins/panel/geomap/layers/data/heatMap.tsx]: :grimacing: ya that looks pretty weird.  I updated the logic to:
```
        const field = weightDim.field ?? frame.fields.find(field => field.type === FieldType.number);
```

File: packages/grafana-ui/src/components/MatchersUI/utils.ts [hunk 3]
```diff
@@ -66,11 +66,15 @@ export function useFieldDisplayNames(data: DataFrame[], filter?: (field: Field)
...
 export function useSelectOptions(
   displayNames: FrameFieldsDisplayNames,
-  currentName?: string
+  currentName?: string,
+  firstItem?: SelectableValue<string>
 ): Array<SelectableValue<string>> {
   return useMemo(() => {
     let found = false;
     const options: Array<SelectableValue<string>> = [];
+    if (firstItem) {
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/MatchersUI/utils.ts:75]: I think `firstItem` has to be also added to the hook's dependency list (there's a terminal warning for that).

File: public/app/plugins/panel/geomap/dims/editors/ScaleDimensionEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+          options={selectOptions}
+          onChange={onSelectChange}
+          noOptionsMessage="No fields found"
+        />
+      </div>
+      <div className={styles.range}>
+        {isFixed && (
+          <InlineFieldRow>
+            <InlineField label="Value" labelWidth={8} grow={true}>
+              <NumberInput value={value.fixed} {...minMaxStep} onChange={onValueChange} />
```
Comment (lean257) [inline public/app/plugins/panel/geomap/dims/editors/ScaleDimensionEditor.tsx:105]: I'm getting a type error `TypeError: Cannot read property 'fixed' of undefined` so it seems that we might want to add optional chaining here `value={value?.fixed}`

**Pull request merged**